rcl/test/rcl/test_wait.cpp

```diff
@@ -41,3 +46,132 @@ TEST(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), test_resize_to_zero) {
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = rcl_wait_set_init(&wait_set, 0, 1, 1, 0, 0, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+
+  // Add a dummy guard condition to avoid an error
+  rcl_guard_condition_t guard_cond = rcl_get_zero_initialized_guard_condition();
+  ret = rcl_guard_condition_init(&guard_cond, rcl_guard_condition_get_default_options());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  ret = rcl_wait_set_add_guard_condition(&wait_set, &guard_cond);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+
+
+  // Time spent during wait should be negligible.
+  int64_t timeout = 0;
+  std::chrono::steady_clock::time_point before_sc = std::chrono::steady_clock::now();
+  ret = rcl_wait(&wait_set, timeout);
+  std::chrono::steady_clock::time_point after_sc = std::chrono::steady_clock::now();
+  // We expect a timeout here (timer value reached)
+  ASSERT_EQ(RCL_RET_TIMEOUT, ret) << rcl_get_error_string_safe();
+  int64_t diff = (after_sc - before_sc).count();
+  EXPECT_LE(diff, TOLERANCE);
+
+  ret = rcl_wait_set_fini(&wait_set);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+}
+
+// Check the interaction of a guard condition and a negative timeout by
+// triggering a guard condition in a separate thread
+TEST(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), guard_condition) {
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = rcl_wait_set_init(&wait_set, 0, 1, 0, 0, 0, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  rcl_guard_condition_t guard_cond = rcl_get_zero_initialized_guard_condition();
+  ret = rcl_guard_condition_init(&guard_cond, rcl_guard_condition_get_default_options());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  ret = rcl_wait_set_add_guard_condition(&wait_set, &guard_cond);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+
+  std::promise<rcl_ret_t> p;
+
+  int64_t timeout = -1;
+
+  std::chrono::nanoseconds trigger_diff;
+  std::thread trigger_thread([&p, &guard_cond, &trigger_diff]() {
+    std::chrono::steady_clock::time_point before_trigger = std::chrono::steady_clock::now();
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    rcl_ret_t ret = rcl_trigger_guard_condition(&guard_cond);
+    std::chrono::steady_clock::time_point after_trigger = std::chrono::steady_clock::now();
+    trigger_diff = std::chrono::duration_cast<std::chrono::nanoseconds>(
+      after_trigger - before_trigger);
+    p.set_value(ret);
+  }
+  );
+  auto f = p.get_future();
+
+  std::chrono::steady_clock::time_point before_sc = std::chrono::steady_clock::now();
+  ret = rcl_wait(&wait_set, timeout);
+  std::chrono::steady_clock::time_point after_sc = std::chrono::steady_clock::now();
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
```

Comment:
Looks like this would cause the guard condition and the wait set to memory leak, but that's probably not a big issue. Nodes and `rcl_init`/`rcl_shutdown` are more important to clean up in exception cases.